sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,219 @@
+    n_bins : int or array-like, shape (n_features_,) (default=2)
+        The number of bins to produce. The intervals for the bins are
+        determined by the minimum and maximum of the input data.
+        Raises ValueError if n_bins < 2.
+
+        If n_bins is an array, and there is a categorical feature at index i,
+        n_bins[i] will be ignored.
+
+    categorical_features : int array-like (default=None)
+        Column indices of categorical features. Categorical features
+        will not be discretized and will be ignored by the transformation.
+
+    Attributes
+    ----------
+    categorical_features_ : int array, shape (n_features_,)
+        Indices of categorical features. These are not affected by the
+        transformation.
+
+    min_ : float array, shape (n_features_,)
+        Minimum value per feature in X. A categorical feature at index i will
+        have min_[i] == 0.
+
+    ptp_ : float array, shape (n_features_,)
+        X.max(axis=0) - X.min(axis=0). A categorical feature at index i will
+        have ptp_[i] == 1.
+
+    n_bins_ : int array, shape (n_features_,)
+        Number of bins per feature. A categorical feature at index i will
+        have n_bins_[i] == 1.
+
+    Example
+    -------
+    >>> X = [[-2, 1, -4,   -1], \
+             [-1, 2, -3, -0.5], \
+             [ 0, 3, -2,  0.5], \
+             [ 1, 4, -1,    2]]
+    >>> dis = KBinsDiscretizer(n_bins=3)
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+    >>> dis.transform(X) # doctest: +SKIP
+    array([[ 0., 0., 0., 0.],
+           [ 1., 1., 1., 0.],
+           [ 2., 2., 2., 1.],
+           [ 2., 2., 2., 2.]])
+
+    Categorical features can be specified in the constructor.
+
+    >>> dis = KBinsDiscretizer(n_bins=3, categorical_features=[1])
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+
+    Discretization will ignore categorical features.
+    >>> dis.transform(X) # doctest: +SKIP
+    array([[ 0., 1., 0., 0.],
+           [ 1., 2., 1., 0.],
+           [ 2., 3., 2., 1.],
+           [ 2., 4., 2., 2.]])
+    """
+
+    def __init__(self, n_bins=2, categorical_features=None):
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+    def fit(self, X, y=None):
+        """Fits the estimator.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            Data to be discretized.
+
+        Returns
+        -------
+        self
+        """
+        X = check_array(X, dtype=float, ensure_2d=False)
+
+        if X.ndim == 1:
+            raise ValueError("Reshape your data.")
+
+        n_features = X.shape[1]
+        self._check_categorical_features(self.categorical_features, X.shape[1])
+        self._check_n_bins(self.n_bins, n_features)
+
+        if self.categorical_features:
+            self.categorical_features_ = np.asarray(self.categorical_features)
```

Comment:
Thanks. There was a time when I was using `self.categorical_features_` in `transform`, but not after the refactoring.